x/ibc/apps/firewall/ibc_middleware.go

```diff
@@ -0,0 +1,233 @@
+package firewall
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	"github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v3/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+	utils "github.com/haqq-network/haqq/types"
+)
+
+var _ porttypes.Middleware = &IBCMiddleware{}
+
+// IBCMiddleware implements the ICS26 callbacks for the fee middleware given the
+// firewall keeper and the underlying application.
+type IBCMiddleware struct {
+	app         porttypes.IBCModule
+	ics4Wrapper porttypes.ICS4Wrapper
+}
+
+// NewIBCMiddleware creates new IBCMiddleware with given ICS4 wrapper and underlying application
+func NewIBCMiddleware(app porttypes.IBCModule, w porttypes.ICS4Wrapper) IBCMiddleware {
+	return IBCMiddleware{
+		app:         app,
+		ics4Wrapper: w,
+	}
+}
+
+// NewICS4Wrapper creates new IBCMiddleware with given underlying ICS4 wrapper
+func NewICS4Wrapper(w porttypes.ICS4Wrapper) IBCMiddleware {
+	return IBCMiddleware{
+		ics4Wrapper: w,
+	}
+}
+
+// IsAllowedAddress checks if given address is allowed to make IBC transfers
```

Comment:
To avoid community confusion, we need to add a comment describing the reason of such restriction.

```go
// Allow only team addresses to ibc-transfer coins outside of the Haqq network.
// 
// This is required until listing happens so that:
// * team is able to transfer ISLM to Gravity Bridge and create a ERC-20 wrap
//   for listing on exchanges which we agreed to list ERC-20 token
// * presale participants and partners won't be able to transfer ISLM
//   to a network with AMMs or DEXes and define a price before the official listing
//
// After the official listing this restriction will be removed.
```

Reply:
Done